### PR TITLE
[ODEInterface] initial BB recipe

### PR DESCRIPTION
**Pull request by luchr**
My second attempt for the initial recipe.

### CODE REVIEW

**Comment by ViralBShah** [comment]
@giordano Shall we merge this?

**Comment by ViralBShah** [comment]
The preferred gcc version is really high. Usually, we want to use the most conservative one we can - unless we need features or optimizations that are only available in newer versions.

**Comment by giordano** [comment]
Isn't there a way to make the recipe shorter?  I think there is a lot of unreasonable duplication here

**Comment by giordano** [comment]
> The preferred gcc version is really high. Usually, we want to use the most conservative one we can - unless we need features or optimizations that are only available in newer versions. Let's see what @giordano has to say about that.

That too (even though when building for libgfortran 3 and 4 we are using older GCCs anyways)

File: O/ODEInterface/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,329 @@
+using BinaryBuilder
+
+name = "ODEInterface"
+version = v"0.0.1"
+
+function get_autogenerated_build_script()
```
Comment (giordano) [inline O/ODEInterface/build_tarballs.jl:6]: Any particular reason why you have a function which returns a constant string?  Can't it just be a string, as usual?

**Comment by giordano** [comment]
>  I think there is a lot of unreasonable duplication here

The difference between all of them seems to be only the extension of the shared library?  Just use [`${dlext}`](https://docs.binarybuilder.org/dev/build_tips/#Automatic-environment-variables)

**Comment by vchuravy** [comment]
This looks like it would benefit from a simple Makefile.

**Comment by luchr** [comment]
@ViralBShah 
> The preferred gcc version is really high. Usually, we want to use the most conservative one we can - unless we need features or optimizations that are only available in newer versions. Let's see what @giordano has to say about that.

No problem. I can change this. One of the fortran files needs `std=f2018` (I already used the deprecated option `std=f2008ts` to also support older compilers).

File: O/ODEInterface/build_tarballs.jl (hunk 1 above)
Comment (luchr) [reply O/ODEInterface/build_tarballs.jl:6]: I can make a static string. I used the function name to try to indicate that this bash-script was auto-generated by the package (`ODEInterface`) itself. I can change this to a comment (or even delete the comment) if this helps.

**Comment by luchr** [comment]
@giordano 
> The difference between all of them seems to be only the extension of the shared library? Just use [`${dlext}`](https://docs.binarybuilder.org/dev/build_tips/#Automatic-environment-variables)

For windows there is e.g. no `-fPIC`.

**Comment by luchr** [comment]
@vchuravy 
> This looks like it would benefit from a simple Makefile.

My goal was to support 3 usage methods:

- User can add the jll-package (manually) and if ODEInterface.jl finds it in the package list, then it tries to use this package.
-  User can call ODEInterface.loadODESolvers with path and then ODEInterface.jl tries to use these libs there (independently how they were produced/compiled/linked).
- User can use the ODEInterface-build-script which tries to build the libs (at the user's host).

And because I already have a build-script ([ODEInterface/deps/build.jl](https://github.com/luchr/ODEInterface.jl/blob/987d0e2cffd1d7f2fd8e378b7292bed4bf451c06/deps/build.jl)) which I need for one of the options above, I wanted to make my life easy and program/update only one/the build-script (it has now the option to execute the commands or to save them; that's the auto-generated bash-script above). If I now have to support/update `build.jl` and a new `Makefile` then this is double-work.

For some reasons I like julia more than bash or the Makefile-syntax and hence I've used this language to describe and execute the building steps (in the past).

**Comment by giordano** [comment]
> For windows there is e.g. no -fPIC.

```
sandbox:${WORKSPACE} # echo $target
x86_64-w64-mingw32
sandbox:${WORKSPACE} # echo 'int foo() {return 0;}' | gcc -shared -fPIC -x c -o foo.dll -
<stdin>:1:0: warning: -fPIC ignored for target (all code is position independent) [enabled by default]
```
You get a warning, but it works

**Comment by luchr** [comment]
@giordano 
> 
> You get a warning, but it works.
I know and that's the reason why I don't use `-fPIC` for windows. 

> 
> Regardless, even if `-fPIC` really didn't work you could do

Please see [my comment above](https://github.com/JuliaPackaging/Yggdrasil/pull/2523#issuecomment-776986004) why I try to only support/update one julia build-script (and use its ugly output for the bash-script thing above).

**Comment by ViralBShah** [comment]
These are solvers that have existed for a long time, probably don't change much at all, and are different projects developed over years. That's why no Makefiles. I think in a future iteration, we can clean it up so that we have OS specific flags and one build script. I say it works, and is pretty good to go - so thanks for merging @giordano!

**Pull request merged**